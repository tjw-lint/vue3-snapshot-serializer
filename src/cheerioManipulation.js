/**
 * @file Uses the Cheerio library to mutate the markup based on the global vueSnapshots settings.
 */

import * as cheerio from 'cheerio';
import * as htmlparser2 from 'htmlparser2';

import {
  stringify,
  swapQuotes
} from './helpers.js';
import { removeTestTokens } from './removeTestTokens.js';

const KEY_NAME = 'data-vue-snapshot-serializer-key';
let key = 0;
let alreadyRemovedKey = true;

/**
 * Safety check to ensure the vueWrapper contains the needed
 * methods for attribute stringification.
 *
 * @param  {object}  vueWrapper  The VTU Wrapper object
 * @return {boolean}             If criteria is met
 */
const attributesCanBeStringified = function (vueWrapper) {
  return (
    (
      globalThis.vueSnapshots?.addInputValues ||
      globalThis.vueSnapshots?.stringifyAttributes
    ) &&
    typeof(vueWrapper?.find) === 'function' &&
    typeof(vueWrapper?.findAll) === 'function'
  );
};

/**
 * Adds a unique key to every DOM element as a data- attribute.
 *
 * @param {object} vueWrapper  The Vue-Test Utils mounted component wrapper
 */
const addSerializerKeys = function (vueWrapper) {
  if (attributesCanBeStringified(vueWrapper)) {
    const vnodes = vueWrapper.findAll('*');
    for (let vnode of vnodes) {
      vnode.element.setAttribute(KEY_NAME, 'v-' + key);
      key++;
    }
    alreadyRemovedKey = false;
  }
};

/**
 * Removes all data-keys from the vueWrapper and Cheerio object.
 *
 * @example
 * <h1 data-vue-snapshot-serializer-key="6">Hello World</h1>
 * <h1>Hello World</h1>
 *
 * @param {object} $           The markup as a cheerio object
 * @param {object} vueWrapper  The Vue-Test Utils mounted component wrapper
 */
const removeSerializerKeys = function ($, vueWrapper) {
  if (!alreadyRemovedKey) {
    $('[' + KEY_NAME + ']').each((index, element) => {
      const currentKey = $(element).attr(KEY_NAME);
      const vnode = vueWrapper.find('[' + KEY_NAME + '="' + currentKey + '"]');
      $(element).removeAttr(KEY_NAME);
      vnode.element.removeAttribute(KEY_NAME);
      alreadyRemovedKey = true;
    });
  }
};

/**
 * Creates a cheerio ($) object from the html for DOM manipulation.
 *
 * @param  {string} html  The markup to use for the cheerio object
 * @return {object}       The cheerio object
 */
const cheerioize = function (html) {
  // https://github.com/fb55/DomHandler
  // https://github.com/fb55/htmlparser2/wiki/Parser-options
  const xmlOptions = {
    decodeEntities: false,
    lowerCaseAttributeNames: false,
    normalizeWhitespace: false,
    recognizeSelfClosing: false,
    xmlMode: false
  };
  const dom = htmlparser2.parseDOM(html, xmlOptions);
  const $ = cheerio.load(dom, { xml: xmlOptions });
  return $;
};

/**
 * Appends a value attribute to input, select, and textareas
 * to show the current value of the element in the snapshot.
 *
 * @example
 * <input>
 * <input value="Hello World">
 *
 * @param {object} $           The markup as a cheerio object
 * @param {object} vueWrapper  The Vue-Test Utils mounted component wrapper
 */
const addInputValues = function ($, vueWrapper) {
  if (
    globalThis.vueSnapshots?.addInputValues &&
    attributesCanBeStringified(vueWrapper)
  ) {
    $('input, textarea, select').each(function (index, element) {
      const currentKey = $(element).attr(KEY_NAME);
      const vnode = vueWrapper.find('[' + KEY_NAME + '="' + currentKey + '"]');
      const value = vnode.element.value;
      element.attribs.value = swapQuotes(stringify(value));
    });
  }
};

/**
 * Replaces dynamic attribute values with a stringified version.
 *
 * @example
 * <h1 title="[object Object]"></h1>
 * <h1 title="{x:'asdf'}"></h1>
 *
 * @param {object} $           The markup as a cheerio object
 * @param {object} vueWrapper  The Vue-Test Utils mounted component wrapper
 */
const stringifyAttributes = function ($, vueWrapper) {
  if (
    globalThis.vueSnapshots?.stringifyAttributes &&
    attributesCanBeStringified(vueWrapper)
  ) {
    $('[' + KEY_NAME + ']').each((index, element) => {
      const currentKey = $(element).attr(KEY_NAME);
      const vnode = vueWrapper.find('[' + KEY_NAME + '="' + currentKey + '"]');
      const attributes = vnode.attributes();
      delete attributes[KEY_NAME];
      const attributeNames = Object.keys(attributes);
      for (let attributeName of attributeNames) {
        let value = vnode?.wrapperElement?.__vnode?.props?.[attributeName];
        if (value !== undefined && typeof(value) !== 'string') {
          value = swapQuotes(stringify(value));
          $(element).attr(attributeName, value);
        }
      }

      // Clean up, remove the serializer data-key
      $(element).removeAttr(KEY_NAME);
      vnode.element.removeAttribute(KEY_NAME);
      alreadyRemovedKey = true;
    });
  }
};

/**
 * This removes data-v-1234abcd="" from your snapshots.
 *
 * @param {object} $  The markup as a cheerio object
 */
const removeScopedStylesDataVIDAttributes = function ($) {
  if (globalThis.vueSnapshots?.removeDataVId) {
    // [-\w]+ will catch 1 or more instaces of a-z, A-Z, 0-9, hyphen (-), or underscore (_)
    const regex = / data-v-[-\w]+/g;

    // [' data-v-asdf=""', ' data-v-qwer', ' data-v-asdf']
    let dataVIds = $.html().match(regex) || [];
    // ['data-v-asdf', 'data-v-qwer', 'data-v-asdf']
    dataVIds = dataVIds.map(function (match) {
      return match.trim().replace('=""', '');
    });
    // ['data-v-asdf', 'data-v-qwer']
    dataVIds = Array.from(new Set(dataVIds));

    dataVIds.forEach(function (attribute) {
      $('[' + attribute + ']').removeAttr(attribute);
    });
  }
};

/**
 * This removes the data-server-rendered="true" from your snapshots.
 *
 * @param {object} $  The markup as a cheerio object
 */
const removeServerRenderedText = function ($) {
  if (globalThis.vueSnapshots?.removeServerRendered) {
    $('[data-server-rendered]').removeAttr('data-server-rendered');
  }
};

/**
 * Loops over the attributesToClear array to set the attribute
 * value to empty string on all matching elements.
 *
 * @param {object} $  The markup as a cheerio object
 */
const clearAttributes = function ($) {
  if (globalThis.vueSnapshots?.attributesToClear?.length) {
    globalThis.vueSnapshots.attributesToClear.forEach(function (attribute) {
      $('[' + attribute + ']').attr(attribute, '');
    });
  }
};

/**
 * Replaces inline functions with the '[function]' placeholder.
 *
 * @param {object} $  The markup as a cheerio object
 */
const clearInlineFunctions = function ($) {
  if (globalThis.vueSnapshots?.clearInlineFunctions) {
    /**
     * Takes a string and tells you if it is a function.
     *
     * @param  {string}  str  Any string
     * @return {boolean}      true = matches function pattern
     */
    const isFunctionDeclaration = function (str) {
      /* eslint-disable-next-line jsdoc/check-line-alignment */
      /**
       * Matches strings that look like functions
       * START:
       *   function
       *   0 or more spaces
       * FUNCTION NAME:
       *   anything 0 or more times
       *   0 or more spaces
       * ARGUMENTS:
       *   (
       *   ARGUMENT:
       *     anything followed by a comma, 0 or more times
       *     0 or more spaces
       *     0 or more times
       *   )
       *   0 or more spaces
       * DECLARATION:
       *   {
       *     maybe anything
       *     maybe return(s)
       *     0 or more times
       *   }
       */
      const functionRegex = /function( )*(.)*( )*\((.,* *){0,}\) *{(.*\n*)*}/;

      if (str.startsWith('function ') || str.startsWith('function(')) {
        return str.endsWith('}') && functionRegex.test(str);
      }

      // Good enough to match most arrow functions
      return /^\s*\w+\s*=>/.test(str) || /^\s*\([^)]*\)\s*=>/.test(str);
    };

    $('*').each(function (index, element) {
      Object.keys(element.attribs).forEach(function (attributeName) {
        let value = element.attribs[attributeName];
        if (isFunctionDeclaration(value)) {
          element.attribs[attributeName] = '[function]';
        }
      });
    });
  }
};

/**
 * Sorts the attributes of all HTML elements to make diffs easier to read.
 *
 * @example
 * <div id="dog" class="cat bat"><h1 title="a" class="b">Text</h1></div>
 * <div class="cat bat" id="dog"><h1 class="b" title="a">Text</h1></div>
 *
 * @param {object} $  The markup as a cheerio object
 */
const sortAttributes = function ($) {
  if (globalThis.vueSnapshots?.sortAttributes) {
    $('*').each(function (index, element) {
      Object.keys(element.attribs).sort().forEach(function (key) {
        let value = element.attribs[key];
        delete element.attribs[key];
        element.attribs[key] = value;
      });
    });
  }
};

/**
 * Applies desired DOM manipulations based on
 * global.vueSnapshots settings for improved snapshots.
 *
 * @param  {object | string} vueWrapper  Either the Vue-Test-Utils mounted component object, or a string of html.
 * @return {string}                      String of manipulated HTML, ready for formatting.
 */
export const cheerioManipulation = function (vueWrapper) {
  addSerializerKeys(vueWrapper);
  let html = vueWrapper;
  if (typeof(vueWrapper?.html) === 'function') {
    html = vueWrapper.html();
  }

  /**
   * NOTE: Although we could check the settings and potentially skip
   * the cheerioze step completely, that would result in inconsistent
   * snapshots, as Cheerio removes empty attribute assignments.
   *
   * `<div class=""></div>` becomes `<div class></div>`
   *
   * Because of this, we should always pass the markup through Cheerio
   * to keep all snapshots consistent, even if we are not doing any
   * DOM manipulation.
   */
  const $ = cheerioize(html);

  addInputValues($, vueWrapper);
  // Removes data-key, so has be last of vueWrapper calls
  stringifyAttributes($, vueWrapper);
  removeServerRenderedText($);
  removeTestTokens($);
  removeScopedStylesDataVIDAttributes($);
  clearAttributes($);
  clearInlineFunctions($);
  sortAttributes($);

  removeSerializerKeys($, vueWrapper);
  return $.html();
};
