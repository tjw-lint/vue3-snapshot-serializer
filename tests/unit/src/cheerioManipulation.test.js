import { mount } from '@vue/test-utils';

import { cheerioManipulation } from '@/cheerioManipulation.js';

import DataVId from '@@/mockComponents/DataVId.vue';
import InlineFunctions from '@@/mockComponents/InlineFunctions.vue';
import SeveralInputs from '@@/mockComponents/SeveralInputs.vue';
import SortAttributes from '@@/mockComponents/SortAttributes.vue';
import StringifyAttributes from '@@/mockComponents/StringifyAttributes.vue';

describe('Cheerio Manipulation', () => {
  beforeEach(() => {
    globalThis.vueSnapshots = {
      formatting: {}
    };
  });

  test('Empty string', () => {
    expect(cheerioManipulation(''))
      .toEqual('');
  });

  describe('Attributes to clear', () => {
    test('Clears attributes', () => {
      globalThis.vueSnapshots.attributesToClear = [
        'data-size',
        'title',
        'non-existent'
      ];

      const markup = [
        '<h1 class="small" data-size="sm" title="text">',
        'Some <strong class="bold" title="Words">text</strong>',
        '</h1>'
      ].join('');

      expect(cheerioManipulation(markup))
        .toEqual([
          '<h1 class="small" data-size title>',
          'Some <strong class="bold" title>text</strong>',
          '</h1>'
        ].join(''));
    });
  });

  describe('Server rendered text', () => {
    const markup = '<div data-server-rendered="true">Content</div>';
    const cleaned = '<div>Content</div>';

    test('Removes server rendered attribute if setting enabled', () => {
      globalThis.vueSnapshots = {
        removeServerRendered: true
      };

      expect(cheerioManipulation(markup))
        .toEqual(cleaned);
    });

    test('Retains server rendered attribute if setting disabled', () => {
      globalThis.vueSnapshots = {
        removeServerRendered: false
      };

      expect(cheerioManipulation(markup))
        .toEqual(markup);
    });
  });

  describe('data-v-ids', () => {
    const markup = '<div data-v-34cd6f4f class="hello"> Hello World </div>';
    const cleaned = '<div class="hello"> Hello World </div>';

    test('Removes data-v-ids from real component', async () => {
      globalThis.vueSnapshots = {
        removeDataVId: true
      };

      const wrapper = await mount(DataVId);

      expect(cheerioManipulation(wrapper.html()))
        .toEqual(cleaned);
    });

    test('Removes data-v-ids from string', () => {
      globalThis.vueSnapshots = {
        removeDataVId: true
      };

      expect(cheerioManipulation(markup))
        .toEqual(cleaned);
    });

    test('Keeps data-v-ids', () => {
      globalThis.vueSnapshots = {
        removeDataVId: false
      };

      expect(cheerioManipulation(markup))
        .toEqual(markup);
    });

    test('Remove empty attributes from data-v-id, but keep the v-id', () => {
      globalThis.vueSnapshots = {
        removeDataVId: false
      };

      const emptyAttribute = '<div data-v-34cd6f4f="" class="hello"> Hello World </div>';

      expect(cheerioManipulation(emptyAttribute))
        .toEqual(markup);
    });
  });

  describe('InlineFunctions.vue', () => {
    test('Functions kept', async () => {
      globalThis.vueSnapshots = {
        clearInlineFunctions: false
      };
      const wrapper = await mount(InlineFunctions);
      const markup = wrapper.html();

      expect(cheerioManipulation(markup))
        .toMatchSnapshot();
    });

    test('Functions removed', async () => {
      globalThis.vueSnapshots = {
        clearInlineFunctions: true
      };
      const wrapper = await mount(InlineFunctions);
      const markup = wrapper.html();

      expect(cheerioManipulation(markup))
        .toMatchSnapshot();
    });

    test('Props', async () => {
      const wrapper = await mount(InlineFunctions);
      const propFn = wrapper.vm.propFn;
      const fn = propFn();

      expect(typeof(propFn))
        .toEqual('function');

      expect(typeof(fn))
        .toEqual('function');

      expect(fn())
        .toEqual({});
    });
  });

  describe('SortAttributes.vue', () => {
    test('Sorted', async () => {
      globalThis.vueSnapshots = { sortAttributes: true };

      const wrapper = await mount(SortAttributes);
      const markup = wrapper.html();

      expect(cheerioManipulation(markup))
        .toMatchSnapshot();
    });

    test('Unsorted', async () => {
      globalThis.vueSnapshots = { sortAttributes: false };

      const wrapper = await mount(SortAttributes);
      const markup = wrapper.html();

      expect(cheerioManipulation(markup))
        .toMatchSnapshot();
    });
  });

  describe('Add input values', () => {
    test('Adds values into DOM', async () => {
      const wrapper = await mount(SeveralInputs);
      await wrapper.find('[data-test="button"]').trigger('click');

      globalThis.vueSnapshots.addInputValues = true;

      expect(wrapper)
        .toMatchSnapshot();
    });

    test('Does not add values into DOM', async () => {
      const wrapper = await mount(SeveralInputs);
      await wrapper.find('[data-test="button"]').trigger('click');

      globalThis.vueSnapshots.addInputValues = false;

      expect(wrapper)
        .toMatchSnapshot();
    });
  });

  describe('Stringify attributes', () => {
    test('Replaces attribute values including child components', async () => {
      globalThis.vueSnapshots.stringifyAttributes = true;

      const wrapper = await mount(StringifyAttributes);

      expect(wrapper)
        .toMatchSnapshot();
    });

    test('Replaces attribute values with stubbed child component', async () => {
      globalThis.vueSnapshots.stringifyAttributes = true;

      const wrapper = await mount(StringifyAttributes, {
        global: {
          stubs: {
            StringifyProps: {
              template: '<span />'
            }
          }
        }
      });

      expect(wrapper)
        .toMatchSnapshot();
    });

    test('Replaces prop values on children in shallow mounts', async () => {
      globalThis.vueSnapshots.stringifyAttributes = true;

      const wrapper = await mount(StringifyAttributes, { shallow: true });

      expect(wrapper)
        .toMatchSnapshot();
    });

    test('Does not stringify attributes', async () => {
      globalThis.vueSnapshots.stringifyAttributes = false;

      const wrapper = await mount(StringifyAttributes);

      expect(wrapper)
        .toMatchSnapshot();
    });
  });
});
