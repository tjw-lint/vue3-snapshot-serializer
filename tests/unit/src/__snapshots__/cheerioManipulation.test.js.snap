// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Cheerio Manipulation > Add input values > Adds values into DOM 1`] = `
<div>
  <input value="'1st'" />
  <textarea value="'second'"></textarea>
  <input
    type="number"
    value="3"
  />
  <select value="'b'">
    <option value="a">
      A
    </option>
    <option value="b">
      B
    </option>
  </select>
  <button>
    Set First
  </button>
</div>
`;

exports[`Cheerio Manipulation > Add input values > Does not add values into DOM 1`] = `
<div>
  <input value="1st" />
  <textarea></textarea>
  <input
    type="number"
    value="3"
  />
  <select>
    <option value="a">
      A
    </option>
    <option value="b">
      B
    </option>
  </select>
  <button>
    Set First
  </button>
</div>
`;

exports[`Cheerio Manipulation > InlineFunctions.vue > Functions kept 1`] = `
<div>
  <h1 title="function () { return true; }">
    Classic
  </h1>
  <h2 title="function named () {
        return true;
      }">
    Multi-line
  </h2>
  <h3 title="(x) => !x">
    Arrow 1
  </h3>
  <h3 title="x => !x">
    Arrow 2
  </h3>
  <h3 title="() => true">
    Arrow 3
  </h3>
  <h3 title="(x, y) => { return x + y; }">
    Arrow 4
  </h3>
  <h4 title="function () { [native code] }">
    Reference
  </h4>
  <h5 title="2">
    Run
  </h5>
  <h6 title="function (a word used in programming) is typically followed by curly braces, ie {}">
    Red herring
  </h6>
  <div>
    <pre>        function notInAnAttribute (a, b, c) {
          const d = a * c;
          const e = d - b;
          return e;
        }
      </pre>
  </div>
  <div title="this is what a it looks like function () {}">
    Attribute does not start with function.
  </div>
  <div title="function () {} is what it looks like">
    Attribute does not end with }.
  </div>
  <p title="function () { [native code] },function () { [native code] }">
    In array
  </p>
  <footer title="function () {
        return () => ({});
      }">
    Default function in a prop
  </footer>
</div>
`;

exports[`Cheerio Manipulation > InlineFunctions.vue > Functions removed 1`] = `
<div>
  <h1 title="[function]">
    Classic
  </h1>
  <h2 title="[function]">
    Multi-line
  </h2>
  <h3 title="[function]">
    Arrow 1
  </h3>
  <h3 title="[function]">
    Arrow 2
  </h3>
  <h3 title="[function]">
    Arrow 3
  </h3>
  <h3 title="[function]">
    Arrow 4
  </h3>
  <h4 title="[function]">
    Reference
  </h4>
  <h5 title="2">
    Run
  </h5>
  <h6 title="function (a word used in programming) is typically followed by curly braces, ie {}">
    Red herring
  </h6>
  <div>
    <pre>        function notInAnAttribute (a, b, c) {
          const d = a * c;
          const e = d - b;
          return e;
        }
      </pre>
  </div>
  <div title="this is what a it looks like function () {}">
    Attribute does not start with function.
  </div>
  <div title="function () {} is what it looks like">
    Attribute does not end with }.
  </div>
  <p title="[function]">
    In array
  </p>
  <footer title="[function]">
    Default function in a prop
  </footer>
</div>
`;

exports[`Cheerio Manipulation > SortAttributes.vue > Sorted 1`] = `
<div>
  <h1
    class="meow nyan"
    id="kitten"
    title="cat"
  >
    fluffy
  </h1>
  <h2
    asdf="qwer"
    class=""
    id="kitten2"
  >
    mittens
  </h2>
  <h3
    asdf=""
    id=""
  >
    scruffy
  </h3>
</div>
`;

exports[`Cheerio Manipulation > SortAttributes.vue > Unsorted 1`] = `
<div>
  <h1
    id="kitten"
    class="meow nyan"
    title="cat"
  >
    fluffy
  </h1>
  <h2
    id="kitten2"
    class=""
    asdf="qwer"
  >
    mittens
  </h2>
  <h3
    id=""
    asdf=""
  >
    scruffy
  </h3>
</div>
`;

exports[`Cheerio Manipulation > Stringify attributes > Does not stringify attributes 1`] = `
<div>
  <h1 title="[object Object]">
    Te
    <s>
      xt
    </s>
  </h1>
</div>
<p title="plain text">
  <strong title="22">
    Text
  </strong>
</p>
<h2 title="[object Object]">
  Words
</h2>
<div title="() => { return 5; }">
  <span title="true,2,string">
    <em>
      Some
    </em>
    stuff.
  </span>
</div>
<div title="[object Object]">
  {
  "name": "bob"
  }
</div>
`;

exports[`Cheerio Manipulation > Stringify attributes > Replaces attribute values including child components 1`] = `
<div>
  <h1 title="{x:'asdf'}">
    Te
    <s>
      xt
    </s>
  </h1>
</div>
<p title="plain text">
  <strong title="22">
    Text
  </strong>
</p>
<h2 title="{y:2}">
  Words
</h2>
<div title="Function">
  <span title="[true,2,'string']">
    <em>
      Some
    </em>
    stuff.
  </span>
</div>
<div title="{name:'bob'}">
  {
  "name": "bob"
  }
</div>
`;

exports[`Cheerio Manipulation > Stringify attributes > Replaces attribute values with stubbed child component 1`] = `
<div>
  <h1 title="{x:'asdf'}">
    Te
    <s>
      xt
    </s>
  </h1>
</div>
<p title="plain text">
  <strong title="22">
    Text
  </strong>
</p>
<h2 title="{y:2}">
  Words
</h2>
<div title="Function">
  <span title="[true,2,'string']">
    <em>
      Some
    </em>
    stuff.
  </span>
</div>
<span person="{name:'bob'}"></span>
`;

exports[`Cheerio Manipulation > Stringify attributes > Replaces prop values on children in shallow mounts 1`] = `
<div>
  <h1 title="{x:'asdf'}">
    Te
    <s>
      xt
    </s>
  </h1>
</div>
<p title="plain text">
  <strong title="22">
    Text
  </strong>
</p>
<h2 title="{y:2}">
  Words
</h2>
<div title="Function">
  <span title="[true,2,'string']">
    <em>
      Some
    </em>
    stuff.
  </span>
</div>
<stringify-props-stub person="{name:'bob'}"></stringify-props-stub>
`;
